build/tools/build_parallel.js

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+'use strict';
+
+var path = require('path');
+var fs = require('fs');
+var findRoot = require('find-root');
+var util = require('util');
+var runAll = require('npm-run-all');
+var async = require('async');
+var EventEmitter = require('events');
+
+// See section titled "About MaxListenersExceededWarning" in https://github.com/mysticatea/npm-run-all/blob/HEAD/docs/node-api.md
+EventEmitter.defaultMaxListeners = 30;
+
+var usage = function() {
+  console.log('usage: node ' + __filename + ' task');
+  console.log('where task is one of:');
+  var tasks = require('./projects.json').tasks;
+  Object.keys(tasks).forEach(function(taskName) {
+    var task = tasks[taskName];
+    console.log("  " + taskName + " - " + task.description);
+  });
+;}
+
+var taskName = (process.argv.length > 2 && process.argv[2]) || 'test';
+var task = require('./projects.json').tasks[taskName];
+if (!task) {
+  usage();
+  process.exit(1);
+}
+
+console.log('dirname is ' + __dirname);
+var gitRoot = findRoot(__dirname, function (dir) {
+  return fs.existsSync(path.resolve(dir, '.gitattributes'));
+});
+
+// ingest projects.json
+var readProjectFile = function() {
+  var projectList = require('./projects.json').projects;
+  var projects = {};
+  projectList.forEach(function (project) {
+    project.fullpath = path.join(gitRoot, project.directory);
+    var packageJson = require(path.join(project.fullpath, 'package.json'));
+    project.dependencies = packageJson.dependencies;
+    project.name = packageJson.name;
+    projects[project.name] = project;
+  });
+  return projects;
+};
+var projects = readProjectFile();
+
+// helper to see if we run the current task for a given project`
+var needToRunTask = function(project) {
+  return (!project.hasOwnProperty(taskName) || !!project[taskName]);
+};
+
+// make a temporary package.json to run our scripts
+var tempDir, packageJsonFilename;
+var makePackageJson = function() {
+  tempDir = path.join(__dirname, 'temp');
+  try {
+    fs.statSync(tempDir).isDirectory();
+  } catch (err) {
+    fs.mkdirSync(tempDir);
+  }
+
+  var packageJson = { 'scripts' : {} };
+
+  Object.keys(projects).forEach(function (projectName) {
+    var project = projects[projectName];
```

Comment:
took me a while to understand what is going on here. please consider 
- documenting/typing the data structures you're using
- revising naming (soo many projectthings in those 2 lines!)
and in this specific case - why not use
```js
for (var project in projects) {
  //...
}
```